### PR TITLE
Add claim airdrop hook app 

### PR DESCRIPTION
**Pull request by JeanNeiverth**
# Summary

- Add Claim Airdrop Hook as a PreHook app;
- Add one example airdrop in order to test the application. The name in the selector is COW, which is only configured for Sepolia network;
- The example follows the ERC20 Merkle Distributor pattern [(repo link)](https://github.com/bleu/cow-airdrop-contract-deployer). The idea is that in order to be compatible with the hook app, future airdrop deployers will need to use this standard.

image

image

image

image

image

image

# To Test

1. Open the page `/#/11155111/swap/hooks/WETH`

- You should be able to see the swap with hooks screen

2. Connect your wallet and switch to Sepolia Network

3. Click 'Add PreHook action' > 'Claim Airdrop' and test the app

4. If your wallet is in [this csv](https://github.com/bleu/cow-airdrop-contract-deployer/blob/example/input-folder/allocations.csv) you should be able to see your claimable tokens. Else, you will probably get the message 'No tokens to claim'

### CODE REVIEW

**Comment by ribeirojose** [comment]
I have read the CLA Document and I hereby sign the CLA

File: apps/cowswap-frontend/src/modules/hooksStore/dapps/AirdropHookApp/airdrop.png
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/hooksStore/dapps/AirdropHookApp/airdrop.png:1]: Awesome icon :)

File: apps/cowswap-frontend/src/modules/hooksStore/dapps/AirdropHookApp/constants.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import { COW } from '@cowprotocol/common-const'
+import { TokenWithLogo } from '@cowprotocol/common-const'
+import { SupportedChainId } from '@cowprotocol/cow-sdk'
+
+import { AirdropOption } from './types'
+
+function buildToken(tokenObj: {
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/hooksStore/dapps/AirdropHookApp/constants.ts:7]: use `TokenWithLogo.fromToken()` instead

File: apps/cowswap-frontend/src/modules/hooksStore/hooks/useGasLimitHooks.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  const provider = useWalletProvider()
+
+  return async (transactionData: Omit<CowHook, 'gasLimit'>) => {
+    if (!provider) throw new Error('Provider is not defined')
+    const gasEstimation = await provider.estimateGas({ to: transactionData.target, data: transactionData.callData })
+    return calculateGasMargin(gasEstimation).toString()
+  }
+}
+
+export const useGasLimit = (to?: string, calldata?: string): string | undefined => {
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/hooksStore/hooks/useGasLimitHooks.ts:22]: Awesome hook! Can we generalise it and move to `@cowprotcol/common-hooks` please?

File: apps/cowswap-frontend/src/modules/hooksStore/hooks/useGasLimitHooks.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  const [gasLimit, setGasLimit] = useState<string>()
+
+  const gasCalculator = useHookGasLimitCalculator()
+
+  useEffect(() => {
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/hooksStore/hooks/useGasLimitHooks.ts:27]: I suggest using `useSWR()` here in order to get better caching results

File: apps/cowswap-frontend/src/modules/hooksStore/dapps/AirdropHookApp/styled/Input.tsx [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { UI } from '@cowprotocol/ui'
+
+import styled from 'styled-components/macro'
+
+export const Input = styled.input`
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/hooksStore/dapps/AirdropHookApp/styled/Input.tsx:5]: We usually create one file `styled.tsx` and put all styled elements there instead of creating a file per element. I would appreciate if you put all of those elements into one file

File: apps/cowswap-frontend/src/modules/hooksStore/dapps/AirdropHookApp/hooks/useClaimData.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  if (airDropData.length < 1) throw new Error(AIRDROP_PREVIEW_ERRORS.NO_CLAIMABLE_AIRDROPS)
+
+  return airDropData[0]
+}
+
+export const useClaimData = (selectedAirdrop?: AirdropOption) => {
+  const { account } = useWalletInfo()
+  const airdropContract = useContract<Airdrop>(selectedAirdrop?.address, AirdropAbi)
+
+  const fetchPreviewClaimableTokens = async ({
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/hooksStore/dapps/AirdropHookApp/hooks/useClaimData.ts:108]: It should be a result of `useCallback()` since it depends on `account` and `airdropContract`

**Comment by alfetopito** [comment]
Build is failing: 
image

File: apps/cowswap-frontend/src/modules/hooksStore/dapps/AirdropHookApp/constants.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  {
+    name: 'COW',
+    dataBaseUrl: 'https://raw.githubusercontent.com/bleu/cow-airdrop-contract-deployer/example/mock-airdrop-data/',
+    chainId: SupportedChainId.SEPOLIA,
+    address: '0xD1fB81659c434DDebC8468713E482134be0D85C0',
+    token: buildToken({
+      ...COW[SupportedChainId.SEPOLIA],
+      address: '0x5fe27bf718937ca1c4a7818d246cd4e755c7470c',
+    }),
+  },
```
[the remark this answers is not in the document]
Comment (yvesfracari) [reply apps/cowswap-frontend/src/modules/hooksStore/dapps/AirdropHookApp/constants.ts:38]: This was our view on the beggining of this hook as well however we thought that wouldn't be too intuitive to add the `databaseUrl` or their data (`index`, `merkle proof`, and `amount`). I know that this way is harder to maintain and request the airdrop team to ask to be added here but we didn't thought a better approach. wdyt? Do you prefer another option?

File: apps/cowswap-frontend/src/modules/hooksStore/hookRegistry.tsx [hunk 7]
```diff
@@ -1,13 +1,14 @@
...
+import { PRE_AIRDROP } from './dapps/AirdropHookApp/hook'
 import { PRE_BUILD, POST_BUILD } from './dapps/BuildHookApp/hook'
 import { PRE_CLAIM_GNO } from './dapps/ClaimGnoHookApp/hook'
 import { HookDapp } from './types/hooks'
 
 export const PRE_HOOK_REGISTRY: Record<SupportedChainId, HookDapp[]> = {
   [SupportedChainId.MAINNET]: [PRE_BUILD],
   [SupportedChainId.GNOSIS_CHAIN]: [PRE_CLAIM_GNO, PRE_BUILD],
-  [SupportedChainId.SEPOLIA]: [PRE_BUILD],
+  [SupportedChainId.SEPOLIA]: [PRE_BUILD, PRE_AIRDROP],
```
[the remark this answers is not in the document]
Comment (yvesfracari) [reply apps/cowswap-frontend/src/modules/hooksStore/hookRegistry.tsx:11]: No reason actually, I think that we can add it!

File: apps/cowswap-frontend/src/modules/hooksStore/dapps/AirdropHookApp/constants.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (alfetopito) [reply apps/cowswap-frontend/src/modules/hooksStore/dapps/AirdropHookApp/constants.ts]: Ok, we talked about this internally and there's no need to make it take any input.

In summary: let's have 1 hook per airdrop. In this case we'll have 1 hook for claiming COW airdrop, another for the test airdrop you are adding.

What would make it easier is to make the base structure for the airdrop generic enough where new hooks can be created with low effort.

**Comment by JeanNeiverth** [comment]
This is how the hook is looking like after removing the airdrop selector

image

**Comment by JeanNeiverth** [comment]
> Build is failing: image

I don't know if it is because we did some changes, but here yarn build is running

image

The yarn build:all triggers some errors, but it doesn't seem it is because anything we changed

image

Can you guys check this again?

**Comment by elena-zh** [comment]
Hey @JeanNeiverth , thank you for clarifications and fixes!

Still, I have some issues to report :) 
1. Edit mode: I don't see 'Return to swap' button there 
image
2. Not eligible to claim: we still display the 'Total available' field. Could you please add 0 COW amount into it, so it will make a bit more sense why this field is displayed on the form? 

3. Not connected wallet: I'd hide this field from the form at all 
not connected
4. Post hooks: if the hook should be displayed there, IMO, it would be great to change its description and remove 'before' from texts (or changed to 'before and/or after')
post - description
post-hook
5. Still, I can see a blurred icon in the hook :( 
blurred
6. 

> Is this OK to add the same hook for the same account several times?
> > I understand that concern, let me circle back to the team to see how we can improve here, but I'm afraid there's no practical way
> >  to avoid that. I know there are plans to work on simulations that will make clear to the user when something's about to go wrong. Do you think this is critical to solve now?

I'm thinking that it would be great to protect users from stupid actions resulting in paying twice for the same transactions. 
@shoom3301 , WDYT? Should this case be fixed here or added as a separate task for addressing later?

7. 
 
> > Might it be possible to show an airdrop amount on the hook details once it is added to the form?
> The situation is pretty similar to 1, we would love to help, but I fear that working on this affects other scopes too. How important do you think it is?;

For me, personally, it is important to show a user an amount to claim and then to swap, if needed. 
@fairlighteth , @shoom3301 , WDYT about this case? 


Thank you!

**Comment by shoom3301** [comment]
>WDYT? Should this case be fixed here or added as a separate task for addressing later?

@elena-zh 
One thing I would like to specify about hooks store - hooks validation and adequacy.
We discussed many times cases like:
what if I add the same hook 3 times?
what if I claim GNO as a prehook, but my sell token will be USDC?
should we allow adding a pre-hook with permitting COW, but selling WETH?
So mainly, those questions are about hooks ordering and hooks influence on the order parameters.
To cover all those questions, I would like to introduce two conceptions.
Hooks are “programming language”. It means, that hooks provide a way to program some actions before and after trade. Using javascript we can write “console.log(1);console.log(1);console.log(1);”. Is this code valid? Yes, it is. Does it make sense? Not so much. Same with hooks, if a hook callData is syntactically valid (Tenderly simulation is successful) - we accept it, but we don’t judge is it logical or not.
Hooks presets / hooks chain / scenarios / etc. This abstraction is responsible for providing a sequence of related on meaning hooks. For example - liquidity migration from Uniswap to CoW AMM. To do this we need: withdraw lp tokens from Uni, swap them, approve spending tokens, transfer them into CoW AMM pool. This scenario must be validated once it’s ready, and it should be validated by a human, I don’t think we can do it programmatically.
In conclusion, there are two independent jobs:
Hooks scenario writing
Hooks scenario usage
First one mainly is supposed to be done by a person with some technical knowledge.
Second one is supposed to be done by any user who interested in its results.

**Comment by shoom3301** [comment]
>Might it be possible to show an airdrop amount on the hook details once it is added to the form?

We have been discussing this with @fairlighteth , it supposed to be displayed as a standardized list of results, like:
```
 - added 30 COW
 - changed recipient to 0x90000...11
 - changed sell token to WETH
```

@elena-zh this is out of this task scope, we will work on it later

**Review by elena-zh — APPROVED** [review]
Thank you for the fixes and explanation.  Approved!

**Comment by alfetopito** [comment]
Now lint is complaining 🙈 
I'll merge as is and address it on our side since it'll be faster, given the GH fork deployment rules.
